README
Calendar-Persian

The README  is  used  to  introduce the module and provide instructions on how to
install the module, any machine dependencies it may have (for example C compilers
and installed libraries) and any other information that should be provided before
the module is installed.

A  README  file  is required for CPAN modules since CPAN extracts the README file
from a module distribution so that people browsing the archive can use it to  get
an  idea  of  the  module's  uses.  It  is usually a good idea to provide version
information here so that people can decide whether fixes for the module are worth
downloading.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After  installing, you  can  find documentation for this module with  the perldoc
command.

    perldoc Calendar::Persian

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Calendar-Persian

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Calendar-Persian

    CPAN Ratings
        http://cpanratings.perl.org/d/Calendar-Persian

    Search CPAN
        http://search.cpan.org/dist/Calendar-Persian/

LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2015 Mohammad S Anwar

This program is free software; you can redistribute it and/or modify it under the
terms of the the Artistic License (2.0).You may obtain a copy of the full license
at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying / distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you, you're nevertheless required to ensure that your Modified Version
complies with the requirements of this license.

This  license  does  not  grant you the right to use any trademark, service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to  make,  have made, use, offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by the  Package.  If you institute patent litigation
(including  a  cross-claim or  counterclaim)  against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF     MERCHANTABILITY,   FITNESS   FOR   A PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE,EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Changes
Revision history for Calendar::Persian

0.34  Wed Jan 11 10:20:00 2017
      - Added dependency on Calendar::Plugin::Renderer v0.12.
      - Removed method validate_params(), no longer required.

0.33  Wed Jan 04 02:10:00 2017
      - Added sample code for method as_text() in the SYNOPSIS section.
      - Updated pod document for method as_text().

0.32  Thu Dec 22 12:05:00 2016
      - Upgraded to use Date::Persian::Simple v0.15.

0.31  Fri Dec 09 14:10:00 2016
      - Added test scripts (t/meta-json.t and t/meta-yml.t).

0.30  Sat Nov 12 21:05:00 2016
      - Upgraded to use Date::Persian::Simple v0.14.

0.29  Tue Oct 18 10:50:00 2016
      - Added dependency on Calendar::Plugin::Renderer v0.11.

0.28  Mon Sep 12 09:50:00 2016
      - Added key 'provides' to the Makefile.PL script.

0.27  Thu Aug 11 11:10:00 2016
      - Updated pod section "SYNOPSIS".
      - Upgraded to use Date::Persian::Simple v0.13.

0.26  Thu Jul 07 07:00:00 2016
      - Added method as_text().
      - Added dependency on Calendar::Plugin::Renderer v0.07.
      - Added test for method as_svg().

0.25  Fri May 27 10:25:00 2016
      - Fixed missing pre-reqs as reported by CPANTS.

0.24  Mon May 23 11:10:00 2016
      - Upgraded to use Date::Persian::Simple v0.10.
      - Updated method as_svg() to handle month name.
      - Refactored method current() and method from_gregorian().
      - Updated pod document with regard to the Calendar::Plugin::Renderer v0.06.

0.23  Fri Apr 29 10:40:00 2016
      - Added method as_svg() using the plugin Calendar::Plugin::Renderer.
      - Upgraded to use Date::Persian::Simple v0.08.
      - Added section "SEE ALSO".
      - Updated copyright year information.

0.22  Mon Aug 31 12:20:00 2015
      - Upgraded to use Date::Persian::Simple v0.07.

0.21  Sun Jul 19 11:15:00 2015
      - Upgraded to use Date::Persian::Simple v0.06.
      - Making use of the method get_calendar() provided by Date::Persian::Simple.

0.20  Thu Jul 09 09:45:00 2015
      - Upgraded to use Date::Persian::Simple v0.05.

0.19  Mon Jun 29 03:25:00 2015
      - Upgraded the dependency on Date::Persian::Simple v0.04.

0.18  Fri Jun 19 10:15:00 2015
      - Fixed the bug in the method from_gregorian().

0.17  Thu Jun 11 10:45:00 2015
      - Updated method _calendar() to validate month and year.

0.16  Mon May 25 13:05:00 2015
      - Dropped the dependency on Date::Utils::Persian.
      - Upgraded the dependency on Date::Persian::Simple v0.03.
      - Updated code w.r.t the above changes.

0.15  Thu Apr 30 12:20:00 2015
      - Added LICENSE file.

0.14  Thu Apr 23 23:40:00 2015
      - Re-structured and consumed Date::Utils::Persian as Moo Role.

0.13  Tue Apr 21 13:40:00 2015
      - Fixed the bug in the method _calendar() the last cell is empty.

0.12  Sun Apr 19 02:50:00 2015
      - Fixed the bug in the method _calendar() when the first day of the month is Sunday.

0.11  Sat Apr 18 03:40:00 2015
      - Complete re-write of the entire distribution using Moo with unit tests.

0.10  Fri Jan 16 15:10:10 2015
      - Removed SIGNATURE file.
      - Updated MANIFEST.SKIP file to ignore folder 'eumm/'.
      - Updated Copyright year information.

0.09  Fri Sep 12 19:42:17 2014
      - Deleted Build.PL file.
      - Removed dependency of Readonly.pm and Carp.pm.
      - Added section "REPOSITORY" to the pod
      - Added key "META_MERGE" to Makefile.PL

0.08  Tue Aug 09 16:25:11 2011
      - Made Readonly v1.03 as minimum requirement.
      - Prepared the distribution using the Module::Signature v0.68.

0.07  Tue Apr 19 15:09:09 2011
      - More cosmetics changes.

0.06  Mon Apr 18 16:30:00 2011
      - Renamed method days_in_year_month() to days_in_month().
      - Renamed title "SYNOPSIS" to "DESCRIPTION".
      - Renamed the title "Persian Calendar for the month of Farvadin year 1390.".
      - Renamed the title "Month Names" to "MONTHS" and set it head1.
      - Set the title "WEEKDAYS" as head1.
      - Minor cosmetics changes.

0.05  Tue Apr 12 12:45:11 2011
      - Fixed the internal method _validate_date().
      - Added unit test 01-test-case.t

0.04  Fri Apr 08 18:58:00 2011
      - Fixed one more typo in the pod for method get_calendar().
      - Removed boilerplate test.

0.03  Fri Apr 08 16:17:18 2011
      - Updated pod for SYNOPSIS.
      - Fixed pod for method days_in_year_month().
      - Fixed pod for method get_calendar().

0.02  Tue Mar 29 10:35:10 2011
      - Added more information about the Persian calendar to the pod.

0.01  Thu Mar 28 17:11:23 2011
      - First version released.

Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'Calendar::Persian',
    AUTHOR             => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    VERSION_FROM       => 'lib/Calendar/Persian.pm',
    ABSTRACT_FROM      => 'lib/Calendar/Persian.pm',
    LICENSE            => 'artistic_2',
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More'        => 0,
        'File::Temp'        => 0,
        'XML::SemanticDiff' => 0,
    },
    PREREQ_PM => {
        'Moo'                        => '2.000000',
        'namespace::clean'           => '0.24',
        'Date::Persian::Simple'      => '0.15',
        'Calendar::Plugin::Renderer' => '0.12',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Calendar-Persian-*' },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        provides    => {
            'Calendar::Persian' => { file => 'lib/Calendar/Persian.pm', version => '0.34' },
        },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/manwar/Calendar-Persian.git',
                web  => 'https://github.com/manwar/Calendar-Persian',
            },
        }})
     : ()
    ),
);


lib/Calendar/Persian.pm
package Calendar::Persian;

$Calendar::Persian::VERSION   = '0.34';
$Calendar::Persian::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

Calendar::Persian - Interface to Persian Calendar.

=head1 VERSION

Version 0.34

=cut

use 5.006;
use Data::Dumper;

use Date::Persian::Simple;
use Moo;
use namespace::clean;
with 'Calendar::Plugin::Renderer';

use overload q{""} => 'as_string', fallback => 1;

has year  => (is => 'rw', predicate => 1);
has month => (is => 'rw', predicate => 1);
has date  => (is => 'ro', default   => sub { Date::Persian::Simple->new });

sub BUILD {
    my ($self) = @_;

    $self->date->validate_year($self->year)   if $self->has_year;
    $self->date->validate_month($self->month) if $self->has_month;

    unless ($self->has_year && $self->has_month) {
        $self->year($self->date->year);
        $self->month($self->date->month);
    }
}

=head1 DESCRIPTION

The Persian  calendar  is  solar, with the particularity that the year defined by
two  successive,  apparent  passages    of  the  Sun  through the vernal (spring)
equinox.  It  is based  on precise astronomical observations, and moreover uses a
sophisticated intercalation system, which makes it more accurate than its younger
European  counterpart,the Gregorian calendar. It is currently used in Iran as the
official  calendar  of  the  country. The  starting  point of the current Iranian
calendar is  the  vernal equinox occurred on Friday March 22 of the year A.D. 622.
Persian Calendar for the month of Farvadin year 1390.

    +---------------------------------------------------------------------------------------------------------------+
    |                                             Farvardin   [1394 BE]                                             |
    +---------------+---------------+---------------+---------------+---------------+---------------+---------------+
    |    Yekshanbeh |     Doshanbeh |    Seshhanbeh | Chaharshanbeh |   Panjshanbeh |         Jomeh |       Shanbeh |
    +---------------+---------------+---------------+---------------+---------------+---------------+---------------+
    |                                                                                               |             1 |
    +---------------+---------------+---------------+---------------+---------------+---------------+---------------+
    |             2 |             3 |             4 |             5 |             6 |             7 |             8 |
    +---------------+---------------+---------------+---------------+---------------+---------------+---------------+
    |             9 |            10 |            11 |            12 |            13 |            14 |            15 |
    +---------------+---------------+---------------+---------------+---------------+---------------+---------------+
    |            16 |            17 |            18 |            19 |            20 |            21 |            22 |
    +---------------+---------------+---------------+---------------+---------------+---------------+---------------+
    |            23 |            24 |            25 |            26 |            27 |            28 |            29 |
    +---------------+---------------+---------------+---------------+---------------+---------------+---------------+
    |            30 |            31 |                                                                               |
    +---------------+---------------+---------------+---------------+---------------+---------------+---------------+

The package L<App::calendr> provides command line tool  C<calendr> to display the
supported calendars on the terminal.

=head1 SYNOPSIS

    use strict; use warnings;
    use Calendar::Persian;

    # prints current month calendar
    print Calendar::Persian->new, "\n";
    print Calendar::Persian->new->current, "\n";

    # prints persian month calendar for the first month of year 1394.
    print Calendar::Persian->new({ month => 1, year => 1394 }), "\n";

    # prints persian month calendar in which the given gregorian date falls in.
    print Calendar::Persian->new->from_gregorian(2015, 1, 14), "\n";

    # prints persian month calendar in which the given julian date falls in.
    print Calendar::Persian->new->from_julian(2457102.5), "\n";

    # prints current month persian calendar in SVG format.
    print Calendar::Persian->new->as_svg;

    # prints current month persian calendar in text format.
    print Calendar::Persian->new->as_text;

=head1 PERSIAN MONTHS

    +-------+-------------------------------------------------------------------+
    | Month | Persian Name                                                      |
    +-------+-------------------------------------------------------------------+
    |     1 | Farvardin                                                         |
    |     2 | Ordibehesht                                                       |
    |     3 | Xordad                                                            |
    |     4 | Tir                                                               |
    |     5 | Amordad                                                           |
    |     6 | Sahrivar                                                          |
    |     7 | Mehr                                                              |
    |     8 | Aban                                                              |
    |     9 | Azar                                                              |
    |    10 | Dey                                                               |
    |    11 | Bahman                                                            |
    |    12 | Esfand                                                            |
    +-------+-------------------------------------------------------------------+

=head1 PERSIAN DAYS

    +-------+---------------+---------------------------------------------------+
    | Index | Persian Name  | English Name                                      |
    +-------+---------------+---------------------------------------------------+
    |     0 | Yekshanbeh    | Sunday                                            |
    |     1 | Doshanbeh     | Monday                                            |
    |     2 | Seshhanbeh    | Tuesday                                           |
    |     3 | Chaharshanbeh | Wednesday                                         |
    |     4 | Panjshanbeh   | Thursday                                          |
    |     5 | Jomeh         | Friday                                            |
    |     6 | Shanbeh       | Saturday                                          |
    +-------+---------------+---------------------------------------------------+

=head1 CONSTRUCTOR

It expects month and year optionally.By default it gets current Persian month and
year.

=head1 METHODS

=head2 current()

Returns current month of the Persian calendar.

=cut

sub current {
    my ($self) = @_;

    return $self->as_text($self->date->month, $self->date->year);
}

=head2 from_gregorian($year, $month, $day)

Returns persian month calendar in which the given gregorian date falls in.

=cut

sub from_gregorian {
    my ($self, $year, $month, $day) = @_;

    return $self->from_julian($self->date->gregorian_to_julian($year, $month, $day));
}

=head2 from_julian($julian_date)

Returns persian month calendar in which the given julian date falls in.

=cut

sub from_julian {
    my ($self, $julian_date) = @_;

    my $date = $self->date->from_julian($julian_date);
    return $self->as_text($date->month, $date->year);
}

=head2 as_svg($month, $year)

Returns calendar for the given C<$month> and C<$year> rendered  in SVG format. If
C<$month> and C<$year> missing, it would return current calendar month.

=cut

sub as_svg {
    my ($self, $month, $year) = @_;

    ($month, $year) = $self->validate_params($month, $year);
    my $date = Date::Persian::Simple->new({ year => $year, month => $month, day => 1 });

    return $self->svg_calendar({
        start_index => $date->day_of_week,
        month_name  => $date->get_month_name,
        days        => $date->days_in_month_year($month, $year),
        year        => $year });
}

=head2 as_text($month, $year)

Returns  color  coded  Persian  calendar for the given C<$month> and C<$year>. If
C<$month> and C<$year> missing, it would return current calendar month.

=cut

sub as_text {
    my ($self, $month, $year) = @_;

    ($month, $year) = $self->validate_params($month, $year);
    my $date = Date::Persian::Simple->new({ year => $year, month => $month, day => 1 });

    return $self->text_calendar(
        {
            start_index => $date->day_of_week,
            month_name  => $date->get_month_name,
            days        => $date->days_in_month_year($month, $year),
            day_names   => $date->days,
            year        => $year
        });
}

sub as_string {
    my ($self) = @_;

    return $self->as_text($self->month, $self->year);
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Calendar-Persian>

=head1 SEE ALSO

=over 4

=item L<Calendar::Bahai>

=item L<Calendar::Hijri>

=item L<Calendar::Gregorian>

=item L<Calendar::Saka>

=back

=head1 BUGS

Please report any bugs or feature requests to C<bug-calendar-persian at rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Calendar-Persian>.
I will be notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Calendar::Persian

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Calendar-Persian>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Calendar-Persian>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Calendar-Persian>

=item * Search CPAN

L<http://search.cpan.org/dist/Calendar-Persian/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2016 Mohammad S Anwar.

This  program  is  free software; you can redistribute it and/or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Calendar::Persian
